### ISSUE TITLE
write_hist_log without log_hist_msec should log job final totals

### ISSUE THREAD

**Issue by malventano** [axboe/fio#490]
There currently appears to be no way to log 'final' histogram data. The closest that is possible is by ensuring log_hist_msec is set to a small enough fraction that it logs near job completion, but this generates excessive and possibly unneeded data where only the single histogram of that particular job was wanted. Further, the last logged histogram will always miss some of the final IOs and will never fully match the final output results (clat percentiles, disk ios, etc) which are totals from the whole job.

Recommendation: Setting write_hist_log=foo without log_hist_msec=x (and/or with log_hist_msec=0) should result in a single histogram line entry produced at the same time as the final output results.

**Comment by malventano** [comment]
After some additional thought, the easiest implementation would be to always write out the histogram bin totals (with write_hist_log set) upon workload completion. This will result in totals logged in absence of log_hist_msec=x and will also result in no missing data, as in the case where log_hist_msec was set, the time between the last interval and the end of the run would be covered by the final total writeout.

**Comment by brycepg** [comment]
`--output-format=json+` causes the `--output` file to contain nanosecond resolution bins for the entire job. Might be what you're looking for

**Comment by bengland2** [comment]
re Nov 12:
If there is only 1 histogram log record and 1 job then you didn't need it at all, since fio calculates percentiles for the entire duration of the job already (though it can't calculate percentiles across set of jobs).

re: nov 28
I tend to agree with this, but what is it actually doing there?  So you're saying that if log_hist_msec=10000 and the test lasts 25 seconds, the last 5 seconds will not be logged?

re: Feb 5
histogram buckets for fio v3 have nanosecond-resolution bins (on the leftmost columns), so I don't see what the difference is.  The question is whether you need nanosecond resolution in the tool that post-processes it, such as fio-histo-log-pctiles.py.

**Comment by malventano** [comment]
(replying to bengland2)
re Nov 12: Histogram records are needed for alternate processing / calculation of non-standard percentiles.
re Nov 28: Correct, the last portion will not be logged. Further, there appears to be no way to get a histogram output of the entire job duration as it is only logged at the requested interval. I do need to retest from the Feb 5 comment to confirm this though.

**Comment by bengland2** [comment]
histogram records are needed for calculation of percentiles wherever there is more than one fio process running.

**Comment by malventano** [comment]
> histogram records are needed for calculation of percentiles wherever there is more than one fio process running.

Agree, but at present it appears that there is no way to get those very histogram records into the log. There is always a gap of IOs at the end of the run, meaning latency percentiles calculated from the histogram logs would not match those in the totals (clat percentiles), as not all IOs were included in the histogram log.

To cover this gap, at run completion, the final histogram bucket counters should be written to the log. This would allow the IOs between the last `log_hist_msec` and the run's end to be covered. It would also allow setting `write_hist_log` without `log_hist_msec` to provide a single set of results (from the entire run), which makes a lot more sense than the blank file we currently get.

**Comment by bengland2** [comment]
I'm no longer working on fio (retired), find someone who is and take it up with them.
